blacklists.py

```diff
@@ -140,17 +147,174 @@ def exists(self, item: Union[str, dict]):
+            (?(DEFINE)(?P<octet>
+              1[0-9]{0,2}|2(?:[0-4][0-9])??|25[0-5]?|2[6-9]|[3-9][0-9]?))
+            ^(?&octet)(?:\.(?&octet)){3}$''', regex.X)
+
+        if 'ip' in item:
+            if not ip_regex.match(item['ip']):
+                raise ValueError('Field "ip" is not a valid IP address: {0}'.format(
+                    item['ip']))
+            '''
+            if 'cidr' in item:
+                raise ValueError(
+                    'Cannot have both "ip" and "cidr" members: {0!r}'.format(item))
+        elif 'cidr' in item:
+            if not 'base' in item['cidr'] or not 'mask' in item['cidr']:
+                raise ValueError('Field "cidr" must have members "base" and "mask"')
+            if not ip_regex.match(item['cidr']['base']):
+                raise ValueError('Field "base" is not a valid IP address: {0}'.format(
+                    item['cidr']['base']))
+            mask = int(item['cidr']['mask'])
+            if mask < 0 or mask > 32:
+                raise ValueError('Field "mask" must be between 0 and 32: {0}'.format(
+                    item['cidr']['mask']))
+            '''
+        else:
+            raise ValueError('Item needs to have an "ip" member field: {0!r}'.format(item))
+
+    def add(self, item):
+        self._validate(item)
+        prikey = self.__class__.SCHEMA_PRIKEY
+
+        def add_callback(d):
+            for compare in d['items']:
+                if compare[prikey] == item[prikey]:
+                    raise ValueError('{0} already in list {1}'.format(
+                        item[prikey], d['items']))
+            d['items'].append(item)
 
+        self._write(add_callback)
+
+    def remove(self, item):
+        prikey = self.__class__.SCHEMA_PRIKEY
+
+        def remove_callback(d):
+            for i, compare in enumerate(d['items']):
+                if compare[prikey] == item[prikey]:
+                    break
+            else:
+                raise ValueError('No {0} found in list {1}'.format(
+                    item[prikey], d['items']))
+            del d['items'][i]
+
+        self._write(remove_callback)
+
+    # FIXME: enumerate gets YAML item array index, not line number
+    def each(self, with_info=False):
+        for i, item in enumerate(self.parse(), start=1):
+            if with_info:
+                yield item, (i, self._filename)
+            else:
+                yield item
+
+    def exists(self, item):
+        item = item.lower()
+        for i, rec in self.each(with_info=True):
+            if item == rec:
+                return True, i
         return False, -1
 
 
+class YAMLParserNS(YAMLParserCIDR):
+    """
+    YAML parser for name server blacklists.
+    """
+    SCHEMA_VARIANT = 'yaml_ns'
+    SCHEMA_PRIKEY = 'ns'
+
+    def _validate(self, item):
+        host_regex = regex.compile(r'^([a-z0-9][-a-z0-9]*\.){2,}$')
+        if 'ns' not in item:
+            raise ValueError('Item must have member field "ns": {0!r}'.format(item))
```

Comment:
Are these ValueErrors caught and logged, or do they crash the instance? If it's the latter, it'd be preferable to change it to the former, even if we have to exit after logging it.

Reply:
Yeah, same as above.